<!DOCTYPE html>
<html lang="zh-CN">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.0.0">


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">



<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5.15.4/css/all.min.css" integrity="sha256-mUZM63G8m73Mcidfrv5E+Y61y7a12O5mW4ezU3bxqW4=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/animate.css@3.1.1/animate.min.css" integrity="sha256-PR7ttpcvz8qrF57fur/yAx1qXMFJeJFiA6pSzWi0OIE=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/pace-js@1.2.4/themes/blue/pace-theme-minimal.css">
  <script src="https://cdn.jsdelivr.net/npm/pace-js@1.2.4/pace.min.js" integrity="sha256-gqd7YTjg/BtfqWSwsJOvndl0Bxc8gFImLEkXQT8+qj0=" crossorigin="anonymous"></script>

<script class="next-config" data-name="main" type="application/json">{"hostname":"yicoyuan.xyz","root":"/","images":"/images","scheme":"Mist","darkmode":false,"version":"8.9.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12},"copycode":false,"bookmark":{"enable":false,"color":"#222","save":"auto"},"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"stickytabs":false,"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"fadeInDown","post_body":"fadeInDown","coll_header":"fadeInLeft","sidebar":"fadeInUp"}},"prism":false,"i18n":{"placeholder":"搜索...","empty":"没有找到任何搜索结果：${query}","hits_time":"找到 ${hits} 个搜索结果（用时 ${time} 毫秒）","hits":"找到 ${hits} 个搜索结果"},"path":"/search.xml","localsearch":{"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false}}</script><script src="/js/config.js"></script>
<meta name="description" content="Pipeline: Mapping→sorting→unique reads(mark duplicates)→indexing">
<meta property="og:type" content="article">
<meta property="og:title" content="测序数据比对到参考基因组">
<meta property="og:url" content="http://yicoyuan.xyz/2022/01/17/WES03/index.html">
<meta property="og:site_name" content="YuanYe&#39;s blog">
<meta property="og:description" content="Pipeline: Mapping→sorting→unique reads(mark duplicates)→indexing">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="http://yicoyuan.xyz/2022/01/17/WES03/01.png">
<meta property="og:image" content="http://yicoyuan.xyz/2022/01/17/WES03/02.png">
<meta property="og:image" content="http://yicoyuan.xyz/2022/01/17/WES03/03.png">
<meta property="og:image" content="http://yicoyuan.xyz/2022/01/17/WES03/04.png">
<meta property="og:image" content="http://yicoyuan.xyz/2022/01/17/WES03/05.png">
<meta property="og:image" content="http://yicoyuan.xyz/2022/01/17/WES03/06.png">
<meta property="article:published_time" content="2022-01-17T12:52:47.000Z">
<meta property="article:modified_time" content="2022-01-28T14:50:33.666Z">
<meta property="article:author" content="Yico Yuan">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://yicoyuan.xyz/2022/01/17/WES03/01.png">


<link rel="canonical" href="http://yicoyuan.xyz/2022/01/17/WES03/">



<script class="next-config" data-name="page" type="application/json">{"sidebar":"","isHome":false,"isPost":true,"lang":"zh-CN","comments":true,"permalink":"http://yicoyuan.xyz/2022/01/17/WES03/","path":"2022/01/17/WES03/","title":"测序数据比对到参考基因组"}</script>

<script class="next-config" data-name="calendar" type="application/json">""</script>
<title>测序数据比对到参考基因组 | YuanYe's blog</title>
  




  <noscript>
    <link rel="stylesheet" href="/css/noscript.css">
  </noscript>
</head>

<body itemscope itemtype="http://schema.org/WebPage" class="use-motion">
  <div class="headband"></div>

  <main class="main">
    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="切换导航栏" role="button">
        <span class="toggle-line"></span>
        <span class="toggle-line"></span>
        <span class="toggle-line"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <i class="logo-line"></i>
      <p class="site-title">YuanYe's blog</p>
      <i class="logo-line"></i>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
        <i class="fa fa-search fa-fw fa-lg"></i>
    </div>
  </div>
</div>



<nav class="site-nav">
  <ul class="main-menu menu">
        <li class="menu-item menu-item-home"><a href="/" rel="section"><i class="fa fa-home fa-fw"></i>首页</a></li>
        <li class="menu-item menu-item-categories"><a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>分类</a></li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>搜索
        </a>
      </li>
  </ul>
</nav>



  <div class="search-pop-overlay">
    <div class="popup search-popup"><div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocapitalize="off" maxlength="80"
           placeholder="搜索..." spellcheck="false"
           type="search" class="search-input">
  </div>
  <span class="popup-btn-close" role="button">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div class="search-result-container no-result">
  <div class="search-result-icon">
    <i class="fa fa-spinner fa-pulse fa-5x"></i>
  </div>
</div>

    </div>
  </div>

</div>
        
  
  <div class="toggle sidebar-toggle" role="button">
    <span class="toggle-line"></span>
    <span class="toggle-line"></span>
    <span class="toggle-line"></span>
  </div>

  <aside class="sidebar">

    <div class="sidebar-inner sidebar-nav-active sidebar-toc-active">
      <ul class="sidebar-nav">
        <li class="sidebar-nav-toc">
          文章目录
        </li>
        <li class="sidebar-nav-overview">
          站点概览
        </li>
      </ul>

      <div class="sidebar-panel-container">
        <!--noindex-->
        <div class="post-toc-wrap sidebar-panel">
            <div class="post-toc animated"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#%E4%B8%BA%E4%BB%80%E4%B9%88%E9%9C%80%E8%A6%81%E5%B0%86%E5%BA%8F%E5%88%97%E6%AF%94%E5%AF%B9%E5%88%B0%E5%8F%82%E8%80%83%E5%9F%BA%E5%9B%A0%E7%BB%84"><span class="nav-number">1.</span> <span class="nav-text">为什么需要将序列比对到参考基因组</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#%E5%8E%9F%E7%90%86"><span class="nav-number">2.</span> <span class="nav-text">原理</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#%E5%B8%B8%E8%A7%81%E6%AF%94%E5%AF%B9%E5%B7%A5%E5%85%B7"><span class="nav-number">3.</span> <span class="nav-text">常见比对工具</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#%E6%AF%94%E5%AF%B9"><span class="nav-number">4.</span> <span class="nav-text">比对</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#%E5%BB%BA%E7%AB%8B%E5%8F%82%E8%80%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E7%B4%A2%E5%BC%95"><span class="nav-number">4.1.</span> <span class="nav-text">建立参考基因组索引</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E6%AF%94%E5%AF%B9-1"><span class="nav-number">4.2.</span> <span class="nav-text">比对</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E8%BE%93%E5%87%BA%E6%AF%94%E5%AF%B9%E4%BF%A1%E6%81%AF"><span class="nav-number">4.3.</span> <span class="nav-text">输出比对信息</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#sam%E6%96%87%E4%BB%B6%E6%8E%92%E5%BA%8F%E5%B9%B6%E8%BD%AC%E6%8D%A2%E4%B8%BAbam"><span class="nav-number">4.4.</span> <span class="nav-text">sam文件排序并转换为bam</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E6%A0%87%E8%AE%B0%E9%87%8D%E5%A4%8D"><span class="nav-number">4.5.</span> <span class="nav-text">标记重复</span></a></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#GATK-resource-bundle"><span class="nav-number">5.</span> <span class="nav-text">GATK resource bundle</span></a></li><li class="nav-item nav-level-1"><a class="nav-link" href="#Prepare-reference-dictionary-fasta-index-and-bam-index"><span class="nav-number">6.</span> <span class="nav-text">Prepare reference dictionary, fasta index, and bam index</span></a></li></ol></div>
        </div>
        <!--/noindex-->

        <div class="site-overview-wrap sidebar-panel">
          <div class="site-author site-overview-item animated" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="Yico Yuan"
      src="/images/ye.svg">
  <p class="site-author-name" itemprop="name">Yico Yuan</p>
  <div class="site-description" itemprop="description">Things To Be Passionate About In Life.</div>
</div>
<div class="site-state-wrap site-overview-item animated">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
        <a href="/archives/">
          <span class="site-state-item-count">9</span>
          <span class="site-state-item-name">日志</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
          <a href="/categories/">
        <span class="site-state-item-count">3</span>
        <span class="site-state-item-name">分类</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author site-overview-item animated">
      <span class="links-of-author-item">
        <a href="mailto:yuanyesulli@outlook.com" title="E-Mail → mailto:yuanyesulli@outlook.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
  </div>



        </div>
      </div>
    </div>
  </aside>
  <div class="sidebar-dimmer"></div>


    </header>

    
  <div class="back-to-top" role="button" aria-label="返回顶部">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>

<noscript>
  <div class="noscript-warning">Theme NexT works best with JavaScript enabled</div>
</noscript>


    <div class="main-inner post posts-expand">


  


<div class="post-block">
  
  

  <article itemscope itemtype="http://schema.org/Article" class="post-content" lang="zh-CN">
    <link itemprop="mainEntityOfPage" href="http://yicoyuan.xyz/2022/01/17/WES03/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/ye.svg">
      <meta itemprop="name" content="Yico Yuan">
      <meta itemprop="description" content="Things To Be Passionate About In Life.">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="YuanYe's blog">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          测序数据比对到参考基因组
        </h1>

        <div class="post-meta-container">
          <div class="post-meta">
    <span class="post-meta-item">
      <span class="post-meta-item-icon">
        <i class="far fa-calendar"></i>
      </span>
      <span class="post-meta-item-text">发表于</span>

      <time title="创建时间：2022-01-17 20:52:47" itemprop="dateCreated datePublished" datetime="2022-01-17T20:52:47+08:00">2022-01-17</time>
    </span>
    <span class="post-meta-item">
      <span class="post-meta-item-icon">
        <i class="far fa-calendar-check"></i>
      </span>
      <span class="post-meta-item-text">更新于</span>
      <time title="修改时间：2022-01-28 22:50:33" itemprop="dateModified" datetime="2022-01-28T22:50:33+08:00">2022-01-28</time>
    </span>
    <span class="post-meta-item">
      <span class="post-meta-item-icon">
        <i class="far fa-folder"></i>
      </span>
      <span class="post-meta-item-text">分类于</span>
        <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
          <a href="/categories/WES/" itemprop="url" rel="index"><span itemprop="name">WES</span></a>
        </span>
    </span>

  
    <span class="post-meta-item" title="阅读次数" id="busuanzi_container_page_pv">
      <span class="post-meta-item-icon">
        <i class="far fa-eye"></i>
      </span>
      <span class="post-meta-item-text">阅读次数：</span>
      <span id="busuanzi_value_page_pv"></span>
    </span>
</div>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">
        <p>Pipeline: Mapping→sorting→unique reads(mark duplicates)→indexing</p>
<span id="more"></span>

<h1 id="为什么需要将序列比对到参考基因组"><a href="#为什么需要将序列比对到参考基因组" class="headerlink" title="为什么需要将序列比对到参考基因组"></a>为什么需要将序列比对到参考基因组</h1><h1 id="原理"><a href="#原理" class="headerlink" title="原理"></a>原理</h1><p>Burrows Wheeler Transformation</p>
<h1 id="常见比对工具"><a href="#常见比对工具" class="headerlink" title="常见比对工具"></a>常见比对工具</h1><p>bwa、bowtie2</p>
<h1 id="比对"><a href="#比对" class="headerlink" title="比对"></a>比对</h1><h2 id="建立参考基因组索引"><a href="#建立参考基因组索引" class="headerlink" title="建立参考基因组索引"></a>建立参考基因组索引</h2><p>绝大多数比对工具都需要先对参考基因组建立索引。我们可以使用下面的命令对参考基因组创建索引文件</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bwa index &lt;reference.fasta&gt;</span><br></pre></td></tr></table></figure>

<p>因此让我们使用这个命令对我们需要的人类参考基因组hg38建立索引：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bwa index ~/WES/ref/hg38.fna</span><br></pre></td></tr></table></figure>

<p>执行成功后，会在参考基因组所在目录生成5个新文件，它们的basename都来自hg38。这些索引文件都是BWA比对需要的。</p>
<p><img src="/2022/01/17/WES03/01.png"></p>
<p>注意：如果参考基因组大于2GB，那么使用BWA建立索引时需要加参数<code>-a bwtsw</code>,这个算法速度慢，但节省内存。</p>
<h2 id="比对-1"><a href="#比对-1" class="headerlink" title="比对"></a>比对</h2><p>建立索引后，我们就可以将测序数据比对到参考基因组上了。用法如下：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bwa mem -M -R &lt;ref&gt; &lt;R1.fastq&gt; &lt;R2.fastq&gt; </span><br></pre></td></tr></table></figure>

<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">参数解释：</span><br><span class="line">mem是bwa比对使用的三种算法之一，支持长序列（70bp~1Mbp）比对,官方推荐的最新算法。</span><br><span class="line">-M 将shorter split hits标记为次优，以兼容Picard’s markDuplicates软件</span><br><span class="line">-R 指定每个read前加上@RG\tID:foo\tSM:bar之类的标头，使用\t分隔。</span><br><span class="line">-t 线程数，默认1</span><br><span class="line">&lt;ref&gt; 参考基因组文件。同时，所有的索引文件都要在同一个文件夹。</span><br><span class="line">&lt;R1.fastq&gt;,&lt;R2.fastq&gt; 单端只输入一个即可。bwa不支持压缩文件，使用记得前先解压。</span><br></pre></td></tr></table></figure>

<p>因此运行以下命令：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">bwa mem -t 8 -M -R &#x27;@RG\tID:sample\tLB:sample\tPL:Ion\tSM:sample&#x27; ~/WES/ref/hg38.fna ~/WES/clean/ERR3013386.clean.fastq \</span><br><span class="line"><span class="meta">&gt;</span><span class="bash"> ~/WES/align/ERR3013386.sam <span class="comment">#使用&gt;将比对结果重定向到.sam文件保存</span></span> </span><br></pre></td></tr></table></figure>

<p>批处理</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">for i in ERR30133&#123;86,87,88,89,91,92,93&#125;</span><br><span class="line">	do</span><br><span class="line">	bwa mem -t 8 -M -R &#x27;@RG\tID:$i\tLB:$i\tPL:Ion\tSM:$i&#x27; \</span><br><span class="line">    ~/WES/ref/hg38.fna \</span><br><span class="line">    ~/WES/clean/$i.clean.fastq \</span><br><span class="line">    &gt; ~/WES/align/$i.sam</span><br><span class="line">done</span><br></pre></td></tr></table></figure>

<h2 id="输出比对信息"><a href="#输出比对信息" class="headerlink" title="输出比对信息"></a>输出比对信息</h2><p>使用samtools flagstat工具对比对结果进行统计：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools flagstat [options] &lt;in.bam&gt;</span><br></pre></td></tr></table></figure>

<p>以单端测序数据ERR3013386的sam比对文件为例：</p>
<p><img src="/2022/01/17/WES03/02.png"></p>
<p>比对结果如下：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">2166159 + 0 in total (QC-passed reads + QC-failed reads)  #reads总数</span><br><span class="line">5038 + 0 secondary  #比对到参考基因组多个位置的reads数</span><br><span class="line">0 + 0 supplementary  #可能存在嵌合的reads数</span><br><span class="line">0 + 0 duplicates  #重复的reads数</span><br><span class="line">2153012 + 0 mapped (99.39% : N/A)  #比对到参考基因组的reads数</span><br><span class="line">0 + 0 paired in sequencing  #属于PE的reads数</span><br><span class="line">0 + 0 read1  #PE中Read_1的reads数</span><br><span class="line">0 + 0 read2  #PE中Read_2的reads数</span><br><span class="line">0 + 0 properly paired (N/A : N/A)  #完美比对的reads数：PE两端reads比对到同一条序列，且根据比对结果推断的插入片段大小符合设置的阈值</span><br><span class="line">0 + 0 with itself and mate mapped  #PE两端都比对上参考序列的reads数</span><br><span class="line">0 + 0 singletons (N/A : N/A)  #PE两端，一端比对上而另一端没比对上的reads数</span><br><span class="line">0 + 0 with mate mapped to a different chr  #PE两端分别比对到不同染色体上的reads数</span><br><span class="line">0 + 0 with mate mapped to a different chr (mapQ&gt;=5)  #PE中两端分别比对到两条不同染色体，且mapQ&gt;=5的reads数</span><br></pre></td></tr></table></figure>

<p>批处理</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">touch ~/WES/align/flagstat.txt #建立一个空txt保存比对结果</span><br><span class="line">for i in ERR30133&#123;86,87,88,89,91,92,93&#125;</span><br><span class="line">	do</span><br><span class="line">	echo $i &gt;&gt; flagstat.txt #追加标头</span><br><span class="line">	samtools flagstat ~/WES/align/$i* &gt;&gt; flagstat.txt #追加比对结果</span><br><span class="line">done</span><br></pre></td></tr></table></figure>

<p>运行完cat一下：</p>
<p><img src="/2022/01/17/WES03/03.png"></p>
<h2 id="sam文件排序并转换为bam"><a href="#sam文件排序并转换为bam" class="headerlink" title="sam文件排序并转换为bam"></a>sam文件排序并转换为bam</h2><p>比对数据的下游分析要求按坐标顺序排好序的bam格式。因此，我们使用samtools工具来操作：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools sort ~/WES/align/ERR3013386.sam -O BAM -o ~/WES/align/ERR3013386.sorted.bam</span><br></pre></td></tr></table></figure>

<p>看一下转换前后有什么不同：</p>
<p><img src="/2022/01/17/WES03/04.png" alt="aligned.sam"></p>
<p><img src="/2022/01/17/WES03/05.png" alt="aligned.sorted.bam"></p>
<p>简单来说就是按照第四列的坐标由小到大重新进行了排列。关于SAM格式文件说明可以详见这篇<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/ab133ee9712c">文章</a>。</p>
<p>批处理：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">for i in ERR30133&#123;86,87,88,89,91,92,93&#125;</span><br><span class="line">  do</span><br><span class="line">  samtools sort ~/WES/align/$i.sam -O BAM -o ~/WES/align/$i.sorted.bam</span><br><span class="line">done</span><br></pre></td></tr></table></figure>

<ul>
<li>比对＋sort</li>
</ul>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">#使用管道避免产生中间的SAM比对文件</span><br><span class="line">for i in ERR30133&#123;86,87,88,89,91,92,93&#125;</span><br><span class="line">	do</span><br><span class="line">	bwa mem -t 8 -M -R &#x27;@RG\tID:$i\tLB:$i\tPL:Ion\tSM:$i&#x27; \</span><br><span class="line">    ~/WES/ref/hg38.fna \</span><br><span class="line">    ~/WES/clean/$i.clean.fastq \</span><br><span class="line">    | samtools sort -O BAM -o ~/WES/align/$i.sorted.bam</span><br><span class="line">done</span><br></pre></td></tr></table></figure>

<h2 id="标记重复"><a href="#标记重复" class="headerlink" title="标记重复"></a>标记重复</h2><p>测序过程中，同一条DNA片段可能会被多次测序。这些重复的reads不是有效信息，也不能作为支持或反对潜在变异的证据。最常见的引起duplicates的原因就是文库构建过程中的PCR。如果不去除这些<strong>PCR duplicates</strong>，PCR扩增偏好性的区域就存在过表达的风险。生成测序cluster时，如果某一个cluster中的DNA序列搭到附近另一个cluster的生成位点，就会又生成一个相同的cluster，产生<strong>Cluster duplicates</strong>。cluster在测序时，捕获的荧光亮点由于光波的衍射，导致形状出现重影，而被当做两个荧光点来处理，从而被认作两条相同的reads，这类称为<strong>Optical duplicates</strong>。</p>
<p>因此我们需要将这些duplicates标记出来，从而得到未被标记的unique reads。最常见的Mark Duplicates工具主要有samtools、picard和sambamba。其中，samtools步骤较繁琐，picard使用最广泛，sambamba处理速度快且用法简单。因此这里推荐使用sambamba。实际上，sambamba开发之初就是为了替代samtools，详细功能参见这篇<a target="_blank" rel="noopener" href="https://github.com/biod/sambamba/wiki/Command-line-tools">文档</a>。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">#用法</span><br><span class="line">sambamba markdup OPTIONS &lt;input.bam&gt; &lt;output.bam&gt;</span><br></pre></td></tr></table></figure>

<p>批处理</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">#这一步会同时建立索引</span><br><span class="line">for i in ERR30133&#123;86,87,88,89,91,92,93&#125;</span><br><span class="line">  do</span><br><span class="line">  sambamba markdup ~/WES/align/$i.sorted.bam $i.sorted.markdup.bam</span><br><span class="line">done</span><br></pre></td></tr></table></figure>

<h1 id="GATK-resource-bundle"><a href="#GATK-resource-bundle" class="headerlink" title="GATK resource bundle"></a>GATK resource bundle</h1><p>下载GATK提供的resource bundle，里面提供了下游分析需要的数据。目前提供FTP 和 Google Cloud bucket 2种下载方式，由于用不了Google，只能使用FTP。GATK官网推荐使用lftp工具进行访问ftp和下载数据。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">#远程访问ftp服务器</span><br><span class="line">lftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/</span><br><span class="line">没有密码，因此直接敲回车即可。</span><br></pre></td></tr></table></figure>

<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ls	#查看以下当前目录下的内容</span><br></pre></td></tr></table></figure>

<p><img src="/2022/01/17/WES03/06.png"></p>
<p>这里我们只需将hg38文件夹里的所有内容下载到本地即可。</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mirror hg38</span><br></pre></td></tr></table></figure>

<p>常见ftp服务器的命令：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">get [OPTS] &lt;rfile&gt; [-o &lt;lfile&gt;] #下载文件</span><br><span class="line">mget [OPTS] &lt;files&gt;  #支持使用通配符来下载文件</span><br><span class="line">mirror [OPTS] [remote [local]]  #下载整个文件夹</span><br><span class="line"></span><br><span class="line">help  #查看所有命令</span><br><span class="line">help &lt;cmd&gt;  #查看某条命令的help</span><br></pre></td></tr></table></figure>

<h1 id="Prepare-reference-dictionary-fasta-index-and-bam-index"><a href="#Prepare-reference-dictionary-fasta-index-and-bam-index" class="headerlink" title="Prepare reference dictionary, fasta index, and bam index"></a>Prepare reference dictionary, fasta index, and bam index</h1><p>使用GATK整合的picard工具建立参考基因组字典文件：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">gatk --java-options &quot;-Xmx8G&quot; CreateSequenceDictionary \</span><br><span class="line">     -R ~/WES/ref/hg38.fna \</span><br><span class="line">     -O ~/WES/ref/hg38.dict</span><br></pre></td></tr></table></figure>

<p>使用samtools建立参考基因组索引：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools faidx ~/WES/ref/hg38.fna  #在参考基因组同位置生成一个.fai索引文件</span><br></pre></td></tr></table></figure>

<p>使用samtools建立markdup.bam文件的索引：</p>
<figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools index markdup.bam  #上面已经生成，这里就简单提一下</span><br></pre></td></tr></table></figure>

<p>这三类索引文件都是GATK下游分析所需要的。其中，前两个文件resource bundle里也有。</p>

    </div>

    
    
    

    <footer class="post-footer">

        

          <div class="post-nav">
            <div class="post-nav-item">
                <a href="/2022/01/15/WES02/" rel="prev" title="测序数据下载及质控">
                  <i class="fa fa-chevron-left"></i> 测序数据下载及质控
                </a>
            </div>
            <div class="post-nav-item">
                <a href="/2022/01/20/WES04/" rel="next" title="碱基质量重矫正">
                  碱基质量重矫正 <i class="fa fa-chevron-right"></i>
                </a>
            </div>
          </div>
    </footer>
  </article>
</div>






</div>
  </main>

  <footer class="footer">
    <div class="footer-inner">


<div class="copyright">
  &copy; 
  <span itemprop="copyrightYear">2022</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Yico Yuan</span>
</div>
<div class="busuanzi-count">
    <span class="post-meta-item" id="busuanzi_container_site_uv">
      <span class="post-meta-item-icon">
        <i class="fa fa-user"></i>
      </span>
      <span class="site-uv" title="总访客量">
        <span id="busuanzi_value_site_uv"></span>
      </span>
    </span>
    <span class="post-meta-item" id="busuanzi_container_site_pv">
      <span class="post-meta-item-icon">
        <i class="fa fa-eye"></i>
      </span>
      <span class="site-pv" title="总访问量">
        <span id="busuanzi_value_site_pv"></span>
      </span>
    </span>
</div>

    </div>
  </footer>

  
  <script src="https://cdn.jsdelivr.net/npm/animejs@3.2.1/lib/anime.min.js" integrity="sha256-XL2inqUJaslATFnHdJOi9GfQ60on8Wx1C2H8DYiN1xY=" crossorigin="anonymous"></script>
<script src="/js/comments.js"></script><script src="/js/utils.js"></script><script src="/js/motion.js"></script><script src="/js/schemes/muse.js"></script><script src="/js/next-boot.js"></script>

  
<script src="https://cdn.jsdelivr.net/npm/hexo-generator-searchdb@1.4.0/dist/search.js" integrity="sha256-vXZMYLEqsROAXkEw93GGIvaB2ab+QW6w3+1ahD9nXXA=" crossorigin="anonymous"></script>
<script src="/js/third-party/search/local-search.js"></script>




  <script src="/js/third-party/pace.js"></script>

  
  <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>





</body>
</html>
