<!DOCTYPE html>
<html lang="zh-CN">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.0.0">


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">



<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5.15.4/css/all.min.css" integrity="sha256-mUZM63G8m73Mcidfrv5E+Y61y7a12O5mW4ezU3bxqW4=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/animate.css@3.1.1/animate.min.css" integrity="sha256-PR7ttpcvz8qrF57fur/yAx1qXMFJeJFiA6pSzWi0OIE=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/pace-js@1.2.4/themes/blue/pace-theme-minimal.css">
  <script src="https://cdn.jsdelivr.net/npm/pace-js@1.2.4/pace.min.js" integrity="sha256-gqd7YTjg/BtfqWSwsJOvndl0Bxc8gFImLEkXQT8+qj0=" crossorigin="anonymous"></script>

<script class="next-config" data-name="main" type="application/json">{"hostname":"yicoyuan.xyz","root":"/","images":"/images","scheme":"Mist","darkmode":false,"version":"8.9.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12},"copycode":false,"bookmark":{"enable":false,"color":"#222","save":"auto"},"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"stickytabs":false,"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"fadeInDown","post_body":"fadeInDown","coll_header":"fadeInLeft","sidebar":"fadeInUp"}},"prism":false,"i18n":{"placeholder":"搜索...","empty":"没有找到任何搜索结果：${query}","hits_time":"找到 ${hits} 个搜索结果（用时 ${time} 毫秒）","hits":"找到 ${hits} 个搜索结果"},"path":"/search.xml","localsearch":{"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false}}</script><script src="/js/config.js"></script>
<meta name="description" content="今天来学习DNA-seq下游分析的第二个内容Filter variants。  Call variants Filter variants Annotation  本篇内容包括了变异质控VQSR建模和hard-filters硬过滤两种不同的方法及原理的介绍。">
<meta property="og:type" content="article">
<meta property="og:title" content="Variants Filtering">
<meta property="og:url" content="http://yicoyuan.xyz/2022/01/25/WES06/index.html">
<meta property="og:site_name" content="YuanYe&#39;s blog">
<meta property="og:description" content="今天来学习DNA-seq下游分析的第二个内容Filter variants。  Call variants Filter variants Annotation  本篇内容包括了变异质控VQSR建模和hard-filters硬过滤两种不同的方法及原理的介绍。">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="http://yicoyuan.xyz/2022/01/25/WES06/01.png">
<meta property="og:image" content="http://yicoyuan.xyz/2022/01/25/WES06/02.png">
<meta property="og:image" content="http://yicoyuan.xyz/2022/01/25/WES06/03.png">
<meta property="article:published_time" content="2022-01-25T15:57:33.000Z">
<meta property="article:modified_time" content="2022-02-12T09:56:04.522Z">
<meta property="article:author" content="Yico Yuan">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://yicoyuan.xyz/2022/01/25/WES06/01.png">


<link rel="canonical" href="http://yicoyuan.xyz/2022/01/25/WES06/">



<script class="next-config" data-name="page" type="application/json">{"sidebar":"","isHome":false,"isPost":true,"lang":"zh-CN","comments":true,"permalink":"http://yicoyuan.xyz/2022/01/25/WES06/","path":"2022/01/25/WES06/","title":"Variants Filtering"}</script>

<script class="next-config" data-name="calendar" type="application/json">""</script>
<title>Variants Filtering | YuanYe's blog</title>
  




  <noscript>
    <link rel="stylesheet" href="/css/noscript.css">
  </noscript>
</head>

<body itemscope itemtype="http://schema.org/WebPage" class="use-motion">
  <div class="headband"></div>

  <main class="main">
    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="切换导航栏" role="button">
        <span class="toggle-line"></span>
        <span class="toggle-line"></span>
        <span class="toggle-line"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <i class="logo-line"></i>
      <p class="site-title">YuanYe's blog</p>
      <i class="logo-line"></i>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
        <i class="fa fa-search fa-fw fa-lg"></i>
    </div>
  </div>
</div>



<nav class="site-nav">
  <ul class="main-menu menu">
        <li class="menu-item menu-item-home"><a href="/" rel="section"><i class="fa fa-home fa-fw"></i>首页</a></li>
        <li class="menu-item menu-item-categories"><a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>分类</a></li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>搜索
        </a>
      </li>
  </ul>
</nav>



  <div class="search-pop-overlay">
    <div class="popup search-popup"><div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocapitalize="off" maxlength="80"
           placeholder="搜索..." spellcheck="false"
           type="search" class="search-input">
  </div>
  <span class="popup-btn-close" role="button">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div class="search-result-container no-result">
  <div class="search-result-icon">
    <i class="fa fa-spinner fa-pulse fa-5x"></i>
  </div>
</div>

    </div>
  </div>

</div>
        
  
  <div class="toggle sidebar-toggle" role="button">
    <span class="toggle-line"></span>
    <span class="toggle-line"></span>
    <span class="toggle-line"></span>
  </div>

  <aside class="sidebar">

    <div class="sidebar-inner sidebar-nav-active sidebar-toc-active">
      <ul class="sidebar-nav">
        <li class="sidebar-nav-toc">
          文章目录
        </li>
        <li class="sidebar-nav-overview">
          站点概览
        </li>
      </ul>

      <div class="sidebar-panel-container">
        <!--noindex-->
        <div class="post-toc-wrap sidebar-panel">
            <div class="post-toc animated"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#VQSR"><span class="nav-number">1.</span> <span class="nav-text">VQSR</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#%E4%BA%BA%E7%B1%BB%E5%8F%98%E5%BC%82%E6%95%B0%E6%8D%AE%E9%9B%86"><span class="nav-number">1.1.</span> <span class="nav-text">人类变异数据集</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#%E8%AE%AD%E7%BB%83%E9%9B%86%E5%8F%82%E6%95%B0%E8%AF%B4%E6%98%8E"><span class="nav-number">1.1.1.</span> <span class="nav-text">训练集参数说明</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#SNP%E5%8F%98%E5%BC%82%E6%95%B0%E6%8D%AE%E9%9B%86"><span class="nav-number">1.1.2.</span> <span class="nav-text">SNP变异数据集</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Indel%E5%8F%98%E5%BC%82%E6%95%B0%E6%8D%AE%E9%9B%86"><span class="nav-number">1.1.3.</span> <span class="nav-text">Indel变异数据集</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#VQSR%E7%9A%84%E5%8E%9F%E7%90%86"><span class="nav-number">1.2.</span> <span class="nav-text">VQSR的原理</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#VQSR%E4%BD%BF%E7%94%A8%E6%96%B9%E6%B3%95"><span class="nav-number">1.3.</span> <span class="nav-text">VQSR使用方法</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#VariantRecalibrator"><span class="nav-number">1.3.1.</span> <span class="nav-text">VariantRecalibrator</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#ApplyVQSR"><span class="nav-number">1.3.2.</span> <span class="nav-text">ApplyVQSR</span></a></li></ol></li></ol></li><li class="nav-item nav-level-1"><a class="nav-link" href="#Hard-filters"><span class="nav-number">2.</span> <span class="nav-text">Hard-filters</span></a></li></ol></div>
        </div>
        <!--/noindex-->

        <div class="site-overview-wrap sidebar-panel">
          <div class="site-author site-overview-item animated" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="Yico Yuan"
      src="/images/ye.svg">
  <p class="site-author-name" itemprop="name">Yico Yuan</p>
  <div class="site-description" itemprop="description">Things To Be Passionate About In Life.</div>
</div>
<div class="site-state-wrap site-overview-item animated">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
        <a href="/archives/">
          <span class="site-state-item-count">9</span>
          <span class="site-state-item-name">日志</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
          <a href="/categories/">
        <span class="site-state-item-count">3</span>
        <span class="site-state-item-name">分类</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author site-overview-item animated">
      <span class="links-of-author-item">
        <a href="mailto:yuanyesulli@outlook.com" title="E-Mail → mailto:yuanyesulli@outlook.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
  </div>



        </div>
      </div>
    </div>
  </aside>
  <div class="sidebar-dimmer"></div>


    </header>

    
  <div class="back-to-top" role="button" aria-label="返回顶部">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>

<noscript>
  <div class="noscript-warning">Theme NexT works best with JavaScript enabled</div>
</noscript>


    <div class="main-inner post posts-expand">


  


<div class="post-block">
  
  

  <article itemscope itemtype="http://schema.org/Article" class="post-content" lang="zh-CN">
    <link itemprop="mainEntityOfPage" href="http://yicoyuan.xyz/2022/01/25/WES06/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/ye.svg">
      <meta itemprop="name" content="Yico Yuan">
      <meta itemprop="description" content="Things To Be Passionate About In Life.">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="YuanYe's blog">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          Variants Filtering
        </h1>

        <div class="post-meta-container">
          <div class="post-meta">
    <span class="post-meta-item">
      <span class="post-meta-item-icon">
        <i class="far fa-calendar"></i>
      </span>
      <span class="post-meta-item-text">发表于</span>

      <time title="创建时间：2022-01-25 23:57:33" itemprop="dateCreated datePublished" datetime="2022-01-25T23:57:33+08:00">2022-01-25</time>
    </span>
    <span class="post-meta-item">
      <span class="post-meta-item-icon">
        <i class="far fa-calendar-check"></i>
      </span>
      <span class="post-meta-item-text">更新于</span>
      <time title="修改时间：2022-02-12 17:56:04" itemprop="dateModified" datetime="2022-02-12T17:56:04+08:00">2022-02-12</time>
    </span>
    <span class="post-meta-item">
      <span class="post-meta-item-icon">
        <i class="far fa-folder"></i>
      </span>
      <span class="post-meta-item-text">分类于</span>
        <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
          <a href="/categories/WES/" itemprop="url" rel="index"><span itemprop="name">WES</span></a>
        </span>
    </span>

  
    <span class="post-meta-item" title="阅读次数" id="busuanzi_container_page_pv">
      <span class="post-meta-item-icon">
        <i class="far fa-eye"></i>
      </span>
      <span class="post-meta-item-text">阅读次数：</span>
      <span id="busuanzi_value_page_pv"></span>
    </span>
</div>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">
        <p>今天来学习DNA-seq下游分析的第二个内容<code>Filter variants</code>。</p>
<ul>
<li>Call variants</li>
<li><strong>Filter variants</strong></li>
<li>Annotation</li>
</ul>
<p>本篇内容包括了变异质控<code>VQSR</code>建模和<code>hard-filters</code>硬过滤两种不同的方法及原理的介绍。</p>
<span id="more"></span>

<h1 id="VQSR"><a href="#VQSR" class="headerlink" title="VQSR"></a>VQSR</h1><p>得到变异数据后，通常需要对其进行过滤质控，筛去假阳性。通常会使用gatk VQSR工具来过滤。</p>
<h2 id="人类变异数据集"><a href="#人类变异数据集" class="headerlink" title="人类变异数据集"></a>人类变异数据集</h2><p>VQSR需要已知的高质量变异数据集作为训练数据进行建模。对于人类而言常见的数据集按质量排序包括SNP: HapMap &gt; Omni &gt; 1000G &gt; dbSNP; Indel: Mills &gt; dbSNP。</p>
<h3 id="训练集参数说明"><a href="#训练集参数说明" class="headerlink" title="训练集参数说明"></a>训练集参数说明</h3><figure class="highlight plaintext"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">--resource:hapmap,known=false,training=true,truth=true,prior=15.0 hapmap_3.3.hg38.sites.vcf.gz</span><br></pre></td></tr></table></figure>

<p>这个<code>--resource</code>指定VQSR用于建模的变异数据集，从左往右用<code>,</code>隔开，依次为：</p>
<ul>
<li>训练集名字，这个名字是可以随便改动的，但是为了便于交流，一般按照数据集的名字来设置。</li>
<li>known：该数据是否作为已知变异数据集，用于对变异数据的标注。</li>
<li>training：该数据是否作为模型训练的数据集，用于训练VQSR模型。</li>
<li>truth：该数据是否作为验证模型训练的真集数据，这个数据同时还是VQSR训练bad model时自动进行参数选择的重要数据。</li>
<li>prior：该数据集在VQSR模型训练中的权重（这里使用Phred值，如20表示0.99）。</li>
</ul>
<h3 id="SNP变异数据集"><a href="#SNP变异数据集" class="headerlink" title="SNP变异数据集"></a>SNP变异数据集</h3><p>HapMap，它来自国际人类单倍体型图计划，HapMap的名字也是源自于此。这个项目刚启动之时，只有270样本，其中有60个家系。项目一共有三期，到第三期HapMap3的时候这个数据已经扩增到1301个样本了，其中有部分样本和千人基因组项目有重叠。由于这个数据集包含了大量家系数据，并且有非常严格的质控和严密的实验验证，因此它的准确性是目前公认最高的。所以VQSR进行质控模型训练的时候，会将其作为一个很重要的训练集（training=true）。它的权重也会被设置得很高，比如在WES数据分析中常常设为prior=15——这里的Prior是Prior likelihood的Phred-scale，我们如果把15转换为likelihood，那么就是0.96838。此外，由于它的高准确性，通常还将作为模型验证的一个真集数据（truth=true）。</p>
<p>Omni，这个数据源自Illumina的Omni基因型芯片，大概2.5百万个位点（我在知识星球中说是1M，这里纠正一下，应该是2.5M），它的验证结果常常作为基因型的金标准。比如用Omni芯片对千人基因组数据进行了验证的结果：1000G_omni2.5.hg38.vcf（这里hg38是参考序列版本），它也是一个高可信的变异结果，我们在VQSR模型训练的时候，同样可以为其设置很高的权重，一般为Prior=12(likelihood为0.9369)。通常情况下也可以把它作为验证结果的真集数据，但我这里所举的例子有些保守，把它设置为truth=false了，大家不必效仿，假如你没太大的“洁癖”把它设置为true都是没问题的（这也是GATK最佳实践的一般做法）。</p>
<p>1000G，千人基因组计划（1000 genomes project）质控后的变异数据，目前也是第三期，一共包含了2504个人的数据。通常来说质控后，它包含的绝大部分都是真实的变异，但由于没办法做全面的实验验证，并不能排除含有少部分假阳的结果。所以模型训练时给的权重虽然比较高——prior=10(likelihood为0.9)，但是一般就不作为模型验证的真集数据了，即truth=false。</p>
<p>dbSNP。说到dbSNP，这是一个绝对不可以作为训练集位点的数据——太脏了，为什么这么说呢？因为，dbSNP收集的数据，实际都是研究者们发表了相关文章提交上来的变异，这些变异很多是没做过严格验证的，很多甚至还是假的，在没被反复验证之前，是不可信的。因此，不会把它们作为模型的训练集，更不会把它作为真集看待（training=false,truth=false），权重也一般设置得很低，比如这里是prior=2（差不多才0.37）。dbSNP的唯一作用就是用于标注我们的变异集中哪些是已经在其它研究中出现过的——即属于已经被发现过的（已知）变异，给这些已知的变异位点标上RS id。</p>
<h3 id="Indel变异数据集"><a href="#Indel变异数据集" class="headerlink" title="Indel变异数据集"></a>Indel变异数据集</h3><p>Mills，对于Indel来说能正在算得上真集的并不多，Mills_and_1000G_gold_standard.indels.hg38.vcf算是其中一个，并被专门做过验证。但其实Indel并不那么容易验证，很多时候也是挑一些比较容易验证的结果。但不管如何，这是目前最佳的一个！所以权重各方面也都设置的比较高，比如prior=12，并且truth=true。</p>
<p>dbSNP，这个就和上面SNP模式下的作用是一样的。不过假如这一步对Indel进行VQSR的VCF数据是顺着上面SNP VQSR后下来的话，那么这个dbSNP的参数可以省略，因为已知变异的标注已经在SNP model下做好了。</p>
<h2 id="VQSR的原理"><a href="#VQSR的原理" class="headerlink" title="VQSR的原理"></a>VQSR的原理</h2><p><img src="/2022/01/25/WES06/01.png" alt="From: 黄树嘉"></p>
<h2 id="VQSR使用方法"><a href="#VQSR使用方法" class="headerlink" title="VQSR使用方法"></a>VQSR使用方法</h2><h3 id="VariantRecalibrator"><a href="#VariantRecalibrator" class="headerlink" title="VariantRecalibrator"></a>VariantRecalibrator</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash">SNP</span></span><br><span class="line">gatk VariantRecalibrator \</span><br><span class="line">   -R ~/WES/ref/hg38.fna \</span><br><span class="line">   -V ~/WES/mutation/raw.vcf \</span><br><span class="line">   --resource:hapmap,known=false,training=true,truth=true,prior=15.0 ~/WES/gatk_resource_bundle/hg38/hapmap_3.3.hg38.ncbi.vcf \</span><br><span class="line">   --resource:omni,known=false,training=true,truth=false,prior=12.0 ~/WES/gatk_resource_bundle/hg38/1000G_omni2.5.hg38.ncbi.vcf \</span><br><span class="line">   --resource:1000G,known=false,training=true,truth=false,prior=10.0 ~/WES/gatk_resource_bundle/hg38/1000G_phase1.snps.high_confidence.hg38.ncbi.vcf \</span><br><span class="line">   --resource:dbsnp,known=true,training=false,truth=false,prior=2.0 ~/WES/gatk_resource_bundle/hg38/dbsnp_146.hg38.ncbi.vcf \</span><br><span class="line">   -an QD -an MQ -an MQRankSum -an ReadPosRankSum -an FS -an SOR \</span><br><span class="line">   --mode SNP \</span><br><span class="line">   -O ~/WES/mutation/VQSR/SNP.recal \</span><br><span class="line">   --tranches-file ~/WES/mutation/VQSR/SNP.tranches \</span><br><span class="line">   --rscript-file ~/WES/mutation/VQSR/SNP.plots.R</span><br><span class="line"><span class="meta"></span></span><br><span class="line"><span class="meta"></span></span><br><span class="line"><span class="meta">#</span><span class="bash">Indel</span></span><br><span class="line">gatk VariantRecalibrator \</span><br><span class="line">   -R ~/WES/ref/hg38.fna \</span><br><span class="line">   -V ~/WES/mutation/raw.vcf \</span><br><span class="line">   --resource:mills,known=false,training=true,truth=true,prior=12.0 ~/WES/gatk_resource_bundle/hg38/Mills_and_1000G_gold_standard.indels.hg38.ncbi.vcf \</span><br><span class="line">   --resource:ddbsnp,known=true,training=false,truth=false,prior=2.0 ~/WES/gatk_resource_bundle/hg38/dbsnp_146.hg38.ncbi.vcf \</span><br><span class="line">   -an QD -an MQ -an MQRankSum -an ReadPosRankSum -an FS -an SOR \</span><br><span class="line">   --mode INDEL \</span><br><span class="line">   -O ~/WES/mutation/VQSR/Indel.recal \</span><br><span class="line">   --tranches-file ~/WES/mutation/VQSR/Indel.tranches \</span><br><span class="line">   --rscript-file ~/WES/mutation/VQSR/Indel.plots.R</span><br></pre></td></tr></table></figure>



<h3 id="ApplyVQSR"><a href="#ApplyVQSR" class="headerlink" title="ApplyVQSR"></a>ApplyVQSR</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash">SNP</span></span><br><span class="line">gatk ApplyVQSR \</span><br><span class="line">   -R ~/WES/ref/hg38.fna \</span><br><span class="line">   -V ~/WES/mutation/raw.vcf \</span><br><span class="line">   -O ~/WES/mutation/VQSR/VQSR.SNP.vcf \</span><br><span class="line">   --truth-sensitivity-filter-level 99.0 \</span><br><span class="line">   --tranches-file ~/WES/mutation/VQSR/SNP.tranches \</span><br><span class="line">   --recal-file ~/WES/mutation/VQSR/SNP.recal \</span><br><span class="line">   --mode SNP</span><br><span class="line"><span class="meta"></span></span><br><span class="line"><span class="meta">#</span><span class="bash">Indel</span></span><br><span class="line">gatk ApplyVQSR \</span><br><span class="line">   -R ~/WES/ref/hg38.fna \</span><br><span class="line">   -V ~/WES/mutation/raw.vcf \</span><br><span class="line">   -O ~/WES/mutation/VQSR/VQSR.Indel.vcf \</span><br><span class="line">   --truth-sensitivity-filter-level 99.0 \</span><br><span class="line">   --tranches-file ~/WES/mutation/VQSR/Indel.tranches \</span><br><span class="line">   --recal-file ~/WES/mutation/VQSR/Indel.recal \</span><br><span class="line">   --mode INDEL</span><br></pre></td></tr></table></figure>

<p>让我们看以下变异位点数量：</p>
<p><img src="/2022/01/25/WES06/02.png"></p>
<p>这个项目本身是外显子测序，再加上样本数量和测序深度都不大，因此数据量本身就小。上图也表面一共就1000个变异，远远达不到训练模型所需要的至少5000个位点的要求。所以不能通过VQSR的方法来对其进行过滤。</p>
<h1 id="Hard-filters"><a href="#Hard-filters" class="headerlink" title="Hard-filters"></a>Hard-filters</h1><p>使用VQSR一是要求我们有可信度高的变异数据集，二是要求我们新测序的结果中有足够多的变异用于模型训练。如果未能达到上述要求，就不得不选择硬过滤的方式，即人为设定若干个指标的阈值，将不满足阈值条件的变异采用一刀切的方式排除掉。通常采取硬过滤使用的是gatk SelectVariants工具，它采用了VQSR所用的6个指标，这些指标都属于VCF的INFO字段（关于VCF文件的说明需要新开一坑），包括：</p>
<ul>
<li>QualByDepth（QD）</li>
<li>FisherStrand (FS)</li>
<li>StrandOddsRatio (SOR)</li>
<li>RMSMappingQuality (MQ)</li>
<li>MappingQualityRankSumTest (MQRankSum)</li>
<li>ReadPosRankSumTest (ReadPosRankSum)</li>
</ul>
<p>那么如何设定合适的阈值呢？</p>
<p>以QualByDepth（QD）为例。首先，什么是QD？QD=Qual/Allele Depth(AD)，即变异质量值除以覆盖深度，即对变异质量的标准化。因为每一个变异的read都对Qual值做出了贡献，测序深度越大其Qual值一般也越大，若不对其进行标准化，质控指标则会因局部深度不均匀而产生偏差。</p>
<p>以下是一个多样本VCF的例子：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">1     1429249 .       C       T       1044.77  .  .  GT:AD:DP:GQ:PL  0/1:48,15:63:99:311,0,1644  0/0:47,0:47:99:392,0,0  1/1:0,76:76:99:3010,228,0</span><br></pre></td></tr></table></figure>

<p>我们可以看到这个变异位点一共有三个样本，其中第一个是杂合变异(GT=0/1)，第二个一个是纯合非变异(GT=0/0)，最后一个是纯合变异(GT=1/1)。其覆盖深度分别为63、47和76。因此这个位点的QD值等于质量值除以另外两个变异样本的覆盖深度：</p>
<p><code>QD=QUAL/AD=1429249/(63+76)=7.516</code></p>
<p>变异位点的QD值越高，其可信度也就越高。那么这个值大于多少，我们才在过滤时判断为真变异呢？简单来说就是一个可信的高深度数据进行VQSR，然后绘出各个指标好坏变异的密度分布图，根据密度分布图来设置阈值。人类的指标阈值就是根据GIAB数据库NA12878的高深度数据进行计算获得的。</p>
<p>最后还会通过计算变异位点Ti/Tv值来检验质控结果。Ti/Tv值是物种在与自然相互作用和演化过程中在基因组上留下来的一个统计标记，具有一定的稳定性。Ti是指嘧啶取代嘧啶或嘌呤取代嘌呤，称为转换；Tv指嘌呤和嘧啶之间的相互取代，称为颠换。如图：</p>
<p><img src="/2022/01/25/WES06/03.png"></p>
<p>对于人来说，全基因组的Ti/Tv在2.1左右，外显子区域在3.0左右，新发生的变异在1.5左右。</p>
<p>详细说明可以参考<a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/34878471">这篇文章</a>。</p>
<p>过滤之前需要先通过gatk SelectVariants的<code>-selectType</code>参数来将VC区分为SNP和Indel这两个不同的变异类型来进行，因为它们有些阈值是不同的，需要区别对待。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">gatk SelectVariants \</span><br><span class="line">	-R ~/WES/ref/hg38.fna \</span><br><span class="line">	-V ~/WES/mutation/raw.vcf \</span><br><span class="line">	-select-type SNP \</span><br><span class="line">	-O ~/WES/mutation/raw_snps.vcf</span><br><span class="line">	</span><br><span class="line">gatk SelectVariants \</span><br><span class="line">	-R ~/WES/ref/hg38.fna \</span><br><span class="line">	-V ~/WES/mutation/raw.vcf \</span><br><span class="line">	-select-type INDEL \</span><br><span class="line">	-O ~/WES/mutation/raw_indels.vcf</span><br></pre></td></tr></table></figure>



<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash">过滤SNP</span></span><br><span class="line">gatk VariantFiltration \</span><br><span class="line">	-R ~/WES/ref/hg38.fna \</span><br><span class="line">	-V ~/WES/mutation/raw_snps.vcf \</span><br><span class="line">	--filter-expression &quot;QD &lt; 2.0 || MQ &lt; 40.0 || FS &gt; 60.0 || SOR &gt; 3.0 || MQRankSum &lt; -12.5 || ReadPosRankSum &lt; -8.0&quot; \</span><br><span class="line">	--filter-name &quot;my_filter&quot; \</span><br><span class="line">	-O ~/WES/mutation/hard_filters/snp.filter.vcf</span><br><span class="line"><span class="meta"></span></span><br><span class="line"><span class="meta">#</span><span class="bash">过滤Indel</span></span><br><span class="line">gatk VariantFiltration \</span><br><span class="line">	-R ~/WES/ref/hg38.fna \</span><br><span class="line">	-V ~/WES/mutation/raw_indels.vcf \</span><br><span class="line">	--filter-expression &quot;QD &lt; 2.0 || FS &gt; 200.0 || SOR &gt; 10.0 || MQRankSum &lt; -12.5 || ReadPosRankSum &lt; -8.0&quot; \</span><br><span class="line">	--filter-name &quot;my_filter&quot; \</span><br><span class="line">	-O ~/WES/mutation/hard_filters/indel.filter.vcf</span><br></pre></td></tr></table></figure>



<p>参考文献：</p>
<ol>
<li><a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/34878471">GATK4.0和全基因组数据分析实践（下）</a></li>
<li><a target="_blank" rel="noopener" href="https://gatk.broadinstitute.org/hc/en-us/articles/4414586833179-VariantRecalibrator">GATK VariantRecalibrator</a></li>
<li><a target="_blank" rel="noopener" href="https://gatk.broadinstitute.org/hc/en-us/articles/4409924805147-ApplyVQSR">GATK ApplyVQSR</a></li>
<li></li>
</ol>

    </div>

    
    
    

    <footer class="post-footer">

        

          <div class="post-nav">
            <div class="post-nav-item">
                <a href="/2022/01/24/WES05/" rel="prev" title="Variants Calling">
                  <i class="fa fa-chevron-left"></i> Variants Calling
                </a>
            </div>
            <div class="post-nav-item">
                <a href="/2022/02/06/WES07/" rel="next" title="变异注释">
                  变异注释 <i class="fa fa-chevron-right"></i>
                </a>
            </div>
          </div>
    </footer>
  </article>
</div>






</div>
  </main>

  <footer class="footer">
    <div class="footer-inner">


<div class="copyright">
  &copy; 
  <span itemprop="copyrightYear">2022</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Yico Yuan</span>
</div>
<div class="busuanzi-count">
    <span class="post-meta-item" id="busuanzi_container_site_uv">
      <span class="post-meta-item-icon">
        <i class="fa fa-user"></i>
      </span>
      <span class="site-uv" title="总访客量">
        <span id="busuanzi_value_site_uv"></span>
      </span>
    </span>
    <span class="post-meta-item" id="busuanzi_container_site_pv">
      <span class="post-meta-item-icon">
        <i class="fa fa-eye"></i>
      </span>
      <span class="site-pv" title="总访问量">
        <span id="busuanzi_value_site_pv"></span>
      </span>
    </span>
</div>

    </div>
  </footer>

  
  <script src="https://cdn.jsdelivr.net/npm/animejs@3.2.1/lib/anime.min.js" integrity="sha256-XL2inqUJaslATFnHdJOi9GfQ60on8Wx1C2H8DYiN1xY=" crossorigin="anonymous"></script>
<script src="/js/comments.js"></script><script src="/js/utils.js"></script><script src="/js/motion.js"></script><script src="/js/schemes/muse.js"></script><script src="/js/next-boot.js"></script>

  
<script src="https://cdn.jsdelivr.net/npm/hexo-generator-searchdb@1.4.0/dist/search.js" integrity="sha256-vXZMYLEqsROAXkEw93GGIvaB2ab+QW6w3+1ahD9nXXA=" crossorigin="anonymous"></script>
<script src="/js/third-party/search/local-search.js"></script>




  <script src="/js/third-party/pace.js"></script>

  
  <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>





</body>
</html>
