<!DOCTYPE html>
<html lang="zh-CN">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 6.0.0">


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">



<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5.15.4/css/all.min.css" integrity="sha256-mUZM63G8m73Mcidfrv5E+Y61y7a12O5mW4ezU3bxqW4=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/animate.css@3.1.1/animate.min.css" integrity="sha256-PR7ttpcvz8qrF57fur/yAx1qXMFJeJFiA6pSzWi0OIE=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/pace-js@1.2.4/themes/blue/pace-theme-minimal.css">
  <script src="https://cdn.jsdelivr.net/npm/pace-js@1.2.4/pace.min.js" integrity="sha256-gqd7YTjg/BtfqWSwsJOvndl0Bxc8gFImLEkXQT8+qj0=" crossorigin="anonymous"></script>

<script class="next-config" data-name="main" type="application/json">{"hostname":"yicoyuan.xyz","root":"/","images":"/images","scheme":"Mist","darkmode":false,"version":"8.9.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12},"copycode":false,"bookmark":{"enable":false,"color":"#222","save":"auto"},"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"stickytabs":false,"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"fadeInDown","post_body":"fadeInDown","coll_header":"fadeInLeft","sidebar":"fadeInUp"}},"prism":false,"i18n":{"placeholder":"搜索...","empty":"没有找到任何搜索结果：${query}","hits_time":"找到 ${hits} 个搜索结果（用时 ${time} 毫秒）","hits":"找到 ${hits} 个搜索结果"},"path":"/search.xml","localsearch":{"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false}}</script><script src="/js/config.js"></script>
<meta name="description" content="GATK是目前业内最权威、使用最广泛的变异检测工具。目前最新版为GATK4，相比GATK3，GATK4在一些工具的使用及参数的设置上都有所优化，取消了Indel的Realignment。">
<meta property="og:type" content="article">
<meta property="og:title" content="碱基质量重矫正">
<meta property="og:url" content="http://yicoyuan.xyz/2022/01/20/WES04/index.html">
<meta property="og:site_name" content="YuanYe&#39;s blog">
<meta property="og:description" content="GATK是目前业内最权威、使用最广泛的变异检测工具。目前最新版为GATK4，相比GATK3，GATK4在一些工具的使用及参数的设置上都有所优化，取消了Indel的Realignment。">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="http://yicoyuan.xyz/2022/01/20/WES04/01.png">
<meta property="og:image" content="http://yicoyuan.xyz/2022/01/20/WES04/02.png">
<meta property="og:image" content="http://yicoyuan.xyz/2022/01/20/WES04/03.png">
<meta property="article:published_time" content="2022-01-20T05:25:27.000Z">
<meta property="article:modified_time" content="2022-01-28T16:19:50.315Z">
<meta property="article:author" content="Yico Yuan">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://yicoyuan.xyz/2022/01/20/WES04/01.png">


<link rel="canonical" href="http://yicoyuan.xyz/2022/01/20/WES04/">



<script class="next-config" data-name="page" type="application/json">{"sidebar":"","isHome":false,"isPost":true,"lang":"zh-CN","comments":true,"permalink":"http://yicoyuan.xyz/2022/01/20/WES04/","path":"2022/01/20/WES04/","title":"碱基质量重矫正"}</script>

<script class="next-config" data-name="calendar" type="application/json">""</script>
<title>碱基质量重矫正 | YuanYe's blog</title>
  




  <noscript>
    <link rel="stylesheet" href="/css/noscript.css">
  </noscript>
</head>

<body itemscope itemtype="http://schema.org/WebPage" class="use-motion">
  <div class="headband"></div>

  <main class="main">
    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="切换导航栏" role="button">
        <span class="toggle-line"></span>
        <span class="toggle-line"></span>
        <span class="toggle-line"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <i class="logo-line"></i>
      <p class="site-title">YuanYe's blog</p>
      <i class="logo-line"></i>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
        <i class="fa fa-search fa-fw fa-lg"></i>
    </div>
  </div>
</div>



<nav class="site-nav">
  <ul class="main-menu menu">
        <li class="menu-item menu-item-home"><a href="/" rel="section"><i class="fa fa-home fa-fw"></i>首页</a></li>
        <li class="menu-item menu-item-categories"><a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>分类</a></li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>搜索
        </a>
      </li>
  </ul>
</nav>



  <div class="search-pop-overlay">
    <div class="popup search-popup"><div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocapitalize="off" maxlength="80"
           placeholder="搜索..." spellcheck="false"
           type="search" class="search-input">
  </div>
  <span class="popup-btn-close" role="button">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div class="search-result-container no-result">
  <div class="search-result-icon">
    <i class="fa fa-spinner fa-pulse fa-5x"></i>
  </div>
</div>

    </div>
  </div>

</div>
        
  
  <div class="toggle sidebar-toggle" role="button">
    <span class="toggle-line"></span>
    <span class="toggle-line"></span>
    <span class="toggle-line"></span>
  </div>

  <aside class="sidebar">

    <div class="sidebar-inner sidebar-nav-active sidebar-toc-active">
      <ul class="sidebar-nav">
        <li class="sidebar-nav-toc">
          文章目录
        </li>
        <li class="sidebar-nav-overview">
          站点概览
        </li>
      </ul>

      <div class="sidebar-panel-container">
        <!--noindex-->
        <div class="post-toc-wrap sidebar-panel">
            <div class="post-toc animated"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#%E7%A2%B1%E5%9F%BA%E8%B4%A8%E9%87%8F%E9%87%8D%E7%9F%AB%E6%AD%A3"><span class="nav-number">1.</span> <span class="nav-text">碱基质量重矫正</span></a><ol class="nav-child"><li class="nav-item nav-level-2"><a class="nav-link" href="#%E5%89%8D%E8%A8%80"><span class="nav-number">1.1.</span> <span class="nav-text">前言</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#BQSR%E8%BF%87%E7%A8%8B%E8%AF%A6%E8%A7%A3"><span class="nav-number">1.2.</span> <span class="nav-text">BQSR过程详解</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#overview"><span class="nav-number">1.2.1.</span> <span class="nav-text">overview</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E7%AC%AC%E4%B8%80%E6%AD%A5%EF%BC%9ABaseRecalibrator%E5%BB%BA%E6%A8%A1"><span class="nav-number">1.2.2.</span> <span class="nav-text">第一步：BaseRecalibrator建模</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E7%AC%AC%E4%BA%8C%E6%AD%A5%EF%BC%9AApplyBQSR-adjusts-the-scores"><span class="nav-number">1.2.3.</span> <span class="nav-text">第二步：ApplyBQSR adjusts the scores</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#BQSR%E5%85%B7%E4%BD%93%E6%93%8D%E4%BD%9C"><span class="nav-number">1.3.</span> <span class="nav-text">BQSR具体操作</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#%E7%AC%AC%E4%B8%80%E6%AD%A5%EF%BC%9A%E5%8D%8F%E5%8F%98%E9%87%8F%E5%88%86%E6%9E%90"><span class="nav-number">1.3.1.</span> <span class="nav-text">第一步：协变量分析</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E7%AC%AC%E4%BA%8C%E6%AD%A5%EF%BC%9A%E8%B0%83%E6%95%B4%E8%B4%A8%E9%87%8F%E5%80%BC"><span class="nav-number">1.3.2.</span> <span class="nav-text">第二步：调整质量值</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E6%89%B9%E5%A4%84%E7%90%86"><span class="nav-number">1.3.3.</span> <span class="nav-text">批处理</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#WHAT-IF"><span class="nav-number">1.3.4.</span> <span class="nav-text">WHAT IF</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#%EF%BC%88%E4%B8%80%EF%BC%89"><span class="nav-number">1.3.4.1.</span> <span class="nav-text">（一）</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#%EF%BC%88%E4%BA%8C%EF%BC%89"><span class="nav-number">1.3.4.2.</span> <span class="nav-text">（二）</span></a></li></ol></li></ol></li></ol></li></ol></div>
        </div>
        <!--/noindex-->

        <div class="site-overview-wrap sidebar-panel">
          <div class="site-author site-overview-item animated" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="Yico Yuan"
      src="/images/ye.svg">
  <p class="site-author-name" itemprop="name">Yico Yuan</p>
  <div class="site-description" itemprop="description">Things To Be Passionate About In Life.</div>
</div>
<div class="site-state-wrap site-overview-item animated">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
        <a href="/archives/">
          <span class="site-state-item-count">11</span>
          <span class="site-state-item-name">日志</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
          <a href="/categories/">
        <span class="site-state-item-count">4</span>
        <span class="site-state-item-name">分类</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author site-overview-item animated">
      <span class="links-of-author-item">
        <a href="mailto:yuanyesulli@outlook.com" title="E-Mail → mailto:yuanyesulli@outlook.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
  </div>



        </div>
      </div>
    </div>
  </aside>
  <div class="sidebar-dimmer"></div>


    </header>

    
  <div class="back-to-top" role="button" aria-label="返回顶部">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>

<noscript>
  <div class="noscript-warning">Theme NexT works best with JavaScript enabled</div>
</noscript>


    <div class="main-inner post posts-expand">


  


<div class="post-block">
  
  

  <article itemscope itemtype="http://schema.org/Article" class="post-content" lang="zh-CN">
    <link itemprop="mainEntityOfPage" href="http://yicoyuan.xyz/2022/01/20/WES04/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/ye.svg">
      <meta itemprop="name" content="Yico Yuan">
      <meta itemprop="description" content="Things To Be Passionate About In Life.">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="YuanYe's blog">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          碱基质量重矫正
        </h1>

        <div class="post-meta-container">
          <div class="post-meta">
    <span class="post-meta-item">
      <span class="post-meta-item-icon">
        <i class="far fa-calendar"></i>
      </span>
      <span class="post-meta-item-text">发表于</span>

      <time title="创建时间：2022-01-20 13:25:27" itemprop="dateCreated datePublished" datetime="2022-01-20T13:25:27+08:00">2022-01-20</time>
    </span>
    <span class="post-meta-item">
      <span class="post-meta-item-icon">
        <i class="far fa-calendar-check"></i>
      </span>
      <span class="post-meta-item-text">更新于</span>
      <time title="修改时间：2022-01-29 00:19:50" itemprop="dateModified" datetime="2022-01-29T00:19:50+08:00">2022-01-29</time>
    </span>
    <span class="post-meta-item">
      <span class="post-meta-item-icon">
        <i class="far fa-folder"></i>
      </span>
      <span class="post-meta-item-text">分类于</span>
        <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
          <a href="/categories/WES/" itemprop="url" rel="index"><span itemprop="name">WES</span></a>
        </span>
    </span>

  
    <span class="post-meta-item" title="阅读次数" id="busuanzi_container_page_pv">
      <span class="post-meta-item-icon">
        <i class="far fa-eye"></i>
      </span>
      <span class="post-meta-item-text">阅读次数：</span>
      <span id="busuanzi_value_page_pv"></span>
    </span>
</div>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">
        <p>GATK是目前业内最权威、使用最广泛的变异检测工具。目前最新版为GATK4，相比GATK3，GATK4在一些工具的使用及参数的设置上都有所优化，取消了Indel的Realignment。</p>
<span id="more"></span>

<p>之所以需要进行Indel局部区域重比对，一是因为BWA和bowtie等全局搜索最优匹配的算法在Indel区域及附近的比对情况往往不是很理想。二是这些算法对于碱基错配和gap的容忍度不同，如果发现碱基错配和gap都可以的情况，会更偏向于错配，这就可能导致基因组上原本应该是一个长Indel的区域被错误地比对为多个错配和短Indel，必然导致检测到错误的变异。</p>
<p>Indel的Realignment就是将BWA比对过程中发现的有潜在Indel的区域进行局部比对，从而重新矫正，降低识别到错误变异的几率。</p>
<p>GATK表示如果使用了最新的HaplotypeCaller或者MuTect2，就不需要这一步了，很可能是GATK4的新工具HaplotypeCaller具备了这一功能。如果使用传统的caller工具，如UnifiedGenotyper或者MuTect，则这一步还是需要的。</p>
<h1 id="碱基质量重矫正"><a href="#碱基质量重矫正" class="headerlink" title="碱基质量重矫正"></a>碱基质量重矫正</h1><h2 id="前言"><a href="#前言" class="headerlink" title="前言"></a>前言</h2><p>BQSR就是Base Quality Score Recalibration，即碱基质量重矫正。这也是数据预处理的最后一步，主要是为了纠正测序仪在测序过程中产生的系统性错误。这类错误有些是因为测序反应中物理或化学反应导致的，有些可能是设备操作上出错了。</p>
<h2 id="BQSR过程详解"><a href="#BQSR过程详解" class="headerlink" title="BQSR过程详解"></a>BQSR过程详解</h2><h3 id="overview"><a href="#overview" class="headerlink" title="overview"></a>overview</h3><p>1.首先提供GATK Base Recalibrator一些列已知的变异集。</p>
<p>2.GATK Base Recalibrator会分析所有的reads，寻找read和reference存在错配的位置，然后排除已知的变异位置。</p>
<p>3.GATK Base Recalibrator根据错配序列碱基原有的质量值、位置和序列内容计算错配碱基的相关数据。</p>
<p>4.基于第3步计算的数据和经验，通过机器学习给每个错配碱基重新计算出新的质量值，代替原来的。</p>
<h3 id="第一步：BaseRecalibrator建模"><a href="#第一步：BaseRecalibrator建模" class="headerlink" title="第一步：BaseRecalibrator建模"></a>第一步：BaseRecalibrator建模</h3><p>为了建立协变量模型，这个工具会遍历输入的BAM文件中所有的reads，然后按列输出碱基的以下特征：</p>
<ul>
<li>read的read group。Read group即前面比对过程中添加以@RG开头的标头，需要通过Read group标签来对reads进行分组矫正。其中：</li>
</ul>
<blockquote>
<p><code>ID</code>=Read group ID，每个read group都有独自的ID。Illumina测序数据中，ID由flowcell、lane name和number组成。矫正碱基质量时，ID对区分技术批次效应是必须的。在这个过程，同一ID的read被假定为有同样的技术误差。</p>
</blockquote>
<blockquote>
<p><code>PU</code>=Platform unit。由三部分组成，<flowcell_barcode>.&lt;lane&gt;.<sample_barcode>。PU不是必须要求的，但是PU的优先级高于ID。</sample_barcode></flowcell_barcode></p>
</blockquote>
<blockquote>
<p><code>SM</code>=Sample。reads所属的样品名。SM要设定正确，因为GATK产生的VCF文件也使用这个名字。</p>
</blockquote>
<blockquote>
<p><code>PL</code>=platform。主要有ILLUMINA、SOLID、LS454、HELICOS和PACBIO等。</p>
</blockquote>
<blockquote>
<p><code>LB</code>=DNA preparation library identifier。对一个read group的reads进行重复序列标记时，需要使用LB来区分reads来自哪条lane。有时同一个库可能在不同的lane上完成的测序，为了加以区分，只要是来自不同的lane都要单独赋予一个ID。</p>
</blockquote>
<ul>
<li>测序仪给出的碱基质量值</li>
<li>碱基产生的轮次（Nth cycle = Nth base from the start of the read）</li>
<li>连续两个碱基（dinucleotide）</li>
</ul>
<p>For each bin, we count the number of bases within the bin and how often such bases mismatch the reference base, excluding loci known to vary in the population, according to the known variants resource (typically dbSNP). This information is output to a recalibration file in GATKReport format.</p>
<p>Note that the recalibrator applies a “yates” correction for low occupancy bins. Rather than inferring the true Q score from # mismatches / # bases we actually infer it from (# mismatches + 1) / (# bases + 2). This deals very nicely with overfitting problems, which has only a minor impact on data sets with billions of bases but is critical to avoid overconfidence in rare bins in sparse data.</p>
<p>这个目前还不是很理解，暂时搬到这里，以后有更深的理解后再更新吧。</p>
<h3 id="第二步：ApplyBQSR-adjusts-the-scores"><a href="#第二步：ApplyBQSR-adjusts-the-scores" class="headerlink" title="第二步：ApplyBQSR adjusts the scores"></a>第二步：ApplyBQSR adjusts the scores</h3><p>This second tool goes through all the reads again, using the recalibration file to adjust each base’s score based on which bins it falls in. So effectively the new quality score is:</p>
<ul>
<li>the sum of the global difference between reported quality scores and the empirical quality</li>
<li>plus the quality bin specific shift</li>
<li>plus the cycle x qual and dinucleotide x qual effect</li>
</ul>
<p>Following recalibration, the read quality scores are much closer to their empirical scores than before. This means they can be used in a statistically robust manner for downstream processing, such as variant calling. In addition, by accounting for quality changes by cycle and sequence context, we can identify truly high quality bases in the reads, often finding a subset of bases that are Q30 even when no bases were originally labeled as such.</p>
<h2 id="BQSR具体操作"><a href="#BQSR具体操作" class="headerlink" title="BQSR具体操作"></a>BQSR具体操作</h2><h3 id="第一步：协变量分析"><a href="#第一步：协变量分析" class="headerlink" title="第一步：协变量分析"></a>第一步：协变量分析</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">gatk BaseRecalibrator \</span><br><span class="line">-R ~/WES/ref/hg38.fna \</span><br><span class="line">-I ~/WES/align/ERR3013386.sorted.markdup.bam \</span><br><span class="line">-known-sites ~/WES/gatk_resource_bundle/hg38/dbsnp_146.hg38.ncbi.vcf \</span><br><span class="line">-known-sites ~/WES/gatk_resource_bundle/hg38/Mills_and_1000G_gold_standard.indels.hg38.ncbi.vcf \</span><br><span class="line">-known-sites ~/WES/gatk_resource_bundle/hg38/1000G_phase1.snps.high_confidence.hg38.ncbi.vcf \</span><br><span class="line">-O ~/WES/align/ERR3013386.recal_data.table</span><br></pre></td></tr></table></figure>





<h3 id="第二步：调整质量值"><a href="#第二步：调整质量值" class="headerlink" title="第二步：调整质量值"></a>第二步：调整质量值</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">gatk ApplyBQSR \</span><br><span class="line">		 -I ~/WES/align/ERR3013386.sorted.markdup.bam \</span><br><span class="line">		 -bqsr ~/WES/align/ERR3013386.recal_data.table \</span><br><span class="line">		 -O ~/WES/align/ERR3013386.markdup.recal.bam</span><br></pre></td></tr></table></figure>

<h3 id="批处理"><a href="#批处理" class="headerlink" title="批处理"></a>批处理</h3><figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">for i in ERR30133&#123;86,87,88,89,91,92,93&#125;</span><br><span class="line">	do</span><br><span class="line">	gatk BaseRecalibrator \</span><br><span class="line">		-R ~/WES/ref/hg38.fna \</span><br><span class="line">		-I ~/WES/align/$i.sorted.markdup.bam \</span><br><span class="line">		-known-sites ~/WES/gatk_resource_bundle/hg38/dbsnp_146.hg38.ncbi.vcf \</span><br><span class="line">		-known-sites ~/WES/gatk_resource_bundle/hg38/Mills_and_1000G_gold_standard.indels.hg38.ncbi.vcf \</span><br><span class="line">		-known-sites ~/WES/gatk_resource_bundle/hg38/1000G_phase1.snps.high_confidence.hg38.ncbi.vcf \</span><br><span class="line">		-O ~/WES/align/$i.recal_data.table </span><br><span class="line">done</span><br><span class="line"></span><br><span class="line">for i in ERR30133&#123;86,87,88,89,91,92,93&#125;</span><br><span class="line">	do</span><br><span class="line">	gatk ApplyBQSR \</span><br><span class="line">		-I ~/WES/align/$i.sorted.markdup.bam \</span><br><span class="line">		-bqsr ~/WES/align/$i.recal_data.table \</span><br><span class="line">		-O ~/WES/align/$i.markdup.recal.bam </span><br><span class="line">done</span><br></pre></td></tr></table></figure>

<p> 关于BQSR要不要使用-L参数来限定仅对外显子区域操作，官方回答是不建议。</p>
<h3 id="WHAT-IF"><a href="#WHAT-IF" class="headerlink" title="WHAT IF"></a>WHAT IF</h3><h4 id="（一）"><a href="#（一）" class="headerlink" title="（一）"></a>（一）</h4><p>如果提示：<code>A USER ERROR has occurred: Input files reference and features have incompatible contigs: No overlapping contigs found.</code>可能是因为从GATK的resource bundle里下载的VCF文件里染色体的命名方式与参考基因组不一致。</p>
<p>NCBI参考基因组文件的标头：<code>NC_000001.11 Homo sapiens chromosome 1, GRCh38.p13 Primary Assembly</code></p>
<p>GRCh38.p13参考基因组NC编号与对应染色体</p>
<table>
<thead>
<tr>
<th align="center">NC_000001.11</th>
<th align="center">NC_000002.12</th>
<th align="center">NC_000003.12</th>
<th align="center">NC_000004.12</th>
<th align="center">NC_000005.10</th>
</tr>
</thead>
<tbody><tr>
<td align="center">chr1</td>
<td align="center">chr2</td>
<td align="center">chr3</td>
<td align="center">chr4</td>
<td align="center">chr5</td>
</tr>
<tr>
<td align="center"><strong>NC_000006.12</strong></td>
<td align="center"><strong>NC_000007.14</strong></td>
<td align="center"><strong>NC_000008.11</strong></td>
<td align="center"><strong>NC_000009.12</strong></td>
<td align="center"><strong>NC_000010.11</strong></td>
</tr>
<tr>
<td align="center">chr6</td>
<td align="center">chr7</td>
<td align="center">chr8</td>
<td align="center">chr9</td>
<td align="center">chr10</td>
</tr>
<tr>
<td align="center"><strong>NC_000011.10</strong></td>
<td align="center"><strong>NC_000012.12</strong></td>
<td align="center"><strong>NC_000013.11</strong></td>
<td align="center"><strong>NC_000014.9</strong></td>
<td align="center"><strong>NC_000015.10</strong></td>
</tr>
<tr>
<td align="center">chr11</td>
<td align="center">chr12</td>
<td align="center">chr13</td>
<td align="center">chr14</td>
<td align="center">chr15</td>
</tr>
<tr>
<td align="center"><strong>NC_000016.10</strong></td>
<td align="center"><strong>NC_000017.11</strong></td>
<td align="center"><strong>NC_000018.10</strong></td>
<td align="center"><strong>NC_000019.10</strong></td>
<td align="center"><strong>NC_000020.11</strong></td>
</tr>
<tr>
<td align="center">chr16</td>
<td align="center">chr17</td>
<td align="center">chr18</td>
<td align="center">chr19</td>
<td align="center">chr20</td>
</tr>
<tr>
<td align="center"><strong>NC_000021.9</strong></td>
<td align="center"><strong>NC_000022.11</strong></td>
<td align="center"><strong>NC_000023.11</strong></td>
<td align="center"><strong>NC_000024.10</strong></td>
<td align="center"><strong>NC_012920.1</strong></td>
</tr>
<tr>
<td align="center">chr21</td>
<td align="center">chr22</td>
<td align="center">chrX</td>
<td align="center">chrY</td>
<td align="center">mitochondrion</td>
</tr>
</tbody></table>
<p>PS:NC编号中小数点后的数字表示版本。</p>
<p>dbsnp.vcf里的染色体命名格式是UCSC风格：<code>chr1    10019   rs775809821     TA      T       .       .       RS=775809821;RSPOS=10020;dbSNPBuildID=144;SSR=0;SAO=0;VP=0x050000020005000002000200;GENEINFO=DDX11L1:100287102;WGT=1;VC=DIV;R5;ASP</code></p>
<p>因此在进行比对时匹配不上。所以解决办法要么改变基因组，要么改变VCF。而改变又有两种方式，要么手动修改，要么下载匹配的格式。</p>
<p>这里选择最头铁的方式，将vcf里UCSC style的染色体名改为NCBI的编号形式。感谢这位兄弟提供的小工具<a target="_blank" rel="noopener" href="https://github.com/vkkodali/cthreepo">cthreepo</a>，支持GFF、GTF、BED、VCF、SAM等文件的染色体ID在NCBI、UCSC和Ensembl之间转换。</p>
<p>安装cthreepo：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash"><span class="built_in">clone</span> the repository</span></span><br><span class="line">git clone https://github.com/vkkodali/cthreepo.git</span><br><span class="line"><span class="meta"></span></span><br><span class="line"><span class="meta">#</span><span class="bash">run the following to install</span></span><br><span class="line">python3 setup.py install</span><br><span class="line"><span class="meta"></span></span><br><span class="line"><span class="meta">#</span><span class="bash">execute as follows</span></span><br><span class="line"><span class="meta">#</span><span class="bash">convert seq-ids <span class="keyword">in</span> &lt;input.gff3&gt; from refseq format (NC_000001.11) to UCSC format (chr1) using the Human GRCh38 mapping dictionary</span></span><br><span class="line">cthreepo \</span><br><span class="line">    --infile &lt;input.gff3&gt; \</span><br><span class="line">    --id_from rs \</span><br><span class="line">    --id_to uc \</span><br><span class="line">    --format gff3 \</span><br><span class="line">    --mapfile h38 \</span><br><span class="line">    --outfile &lt;output.gff3&gt;</span><br><span class="line">    </span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="built_in">help</span></span></span><br><span class="line">cthreepo --help</span><br></pre></td></tr></table></figure>

<p>然后下载NCBI提供的<a target="_blank" rel="noopener" href="https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.39_GRCh38.p13/GCF_000001405.39_GRCh38.p13_assembly_report.txt">assembly report</a>文件到本地或者添加-a参数指定NCBI Assembly Accession with version来在线获取，cthreepo就是根据这个文件的ID注释进行转换的。</p>
<p>将3个vcf文件转换为我们所需要的：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash">Mills_and_1000G</span></span><br><span class="line">cthreepo \</span><br><span class="line">--infile ~/WES/gatk_resource_bundle/hg38/Mills_and_1000G_gold_standard.indels.hg38.vcf \</span><br><span class="line">--id_from uc --id_to rs --format vcf \</span><br><span class="line">-a GCF_000001405.39 \</span><br><span class="line">--outfile ~/WES/gatk_resource_bundle/hg38/Mills_and_1000G_gold_standard.indels.hg38.ncbi.vcf</span><br><span class="line"><span class="meta"></span></span><br><span class="line"><span class="meta">#</span><span class="bash">1000G</span></span><br><span class="line">cthreepo \</span><br><span class="line">--infile ~/WES/gatk_resource_bundle/hg38/1000G_phase1.snps.high_confidence.hg38.vcf \</span><br><span class="line">--id_from uc --id_to rs --format vcf \</span><br><span class="line">-a GCF_000001405.39 \</span><br><span class="line">--outfile ~/WES/gatk_resource_bundle/hg38/1000G_phase1.snps.high_confidence.hg38.ncbi.vcf</span><br><span class="line"><span class="meta"></span></span><br><span class="line"><span class="meta">#</span><span class="bash">dbsnp_146</span></span><br><span class="line">cthreepo \</span><br><span class="line">--infile ~/WES/gatk_resource_bundle/hg38/dbsnp_146.hg38.vcf \</span><br><span class="line">--id_from uc --id_to rs --format vcf \</span><br><span class="line">-a GCF_000001405.39 \</span><br><span class="line">--outfile ~/WES/gatk_resource_bundle/hg38/dbsnp_146.hg38.ncbi.vcf</span><br></pre></td></tr></table></figure>

<p>看看转换前后对比：</p>
<p><img src="/2022/01/20/WES04/01.png" alt="before"></p>
<p><img src="/2022/01/20/WES04/02.png" alt="after"></p>
<p>但是再次运行gatk BaseRecalibrator后发现还是提示reference和features的contigs还是不匹配。单独指定vcf后运行发现，仅dbsnp_146运行成功。通过比较与另外两个VCF的差别后发现，dbsnp_146没有<code>##contig=&lt; &gt;</code>这部分codecs：</p>
<p><img src="/2022/01/20/WES04/03.png"></p>
<p>因此猜想是这部分编码内容误导了gatk，于是将这两个VCF里的删去，并重建索引：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash">以下代码在文件所在目录运行</span></span><br><span class="line">mv Mills_and_1000G_gold_standard.indels.hg38.ncbi.vcf Mills_and_1000G_gold_standard.indels.hg38.ncbi.old.vcf</span><br><span class="line">grep -v &quot;##contig=&quot; Mills_and_1000G_gold_standard.indels.hg38.ncbi.old.vcf &gt; Mills_and_1000G_gold_standard.indels.hg38.ncbi.vcf </span><br><span class="line">gatk IndexFeatureFile -I Mills_and_1000G_gold_standard.indels.hg38.ncbi.vcf</span><br><span class="line">rm -rf Mills_and_1000G_gold_standard.indels.hg38.ncbi.old.vcf</span><br><span class="line"><span class="meta"></span></span><br><span class="line"><span class="meta">#</span><span class="bash">另一个同理</span></span><br><span class="line">mv 1000G_phase1.snps.high_confidence.hg38.ncbi.vcf 1000G_phase1.snps.high_confidence.hg38.ncbi.old.vcf</span><br><span class="line">grep -v &quot;##contig=&quot; 1000G_phase1.snps.high_confidence.hg38.ncbi.old.vcf &gt; 1000G_phase1.snps.high_confidence.hg38.ncbi.vcf </span><br><span class="line">gatk IndexFeatureFile -I 1000G_phase1.snps.high_confidence.hg38.ncbi.vcf</span><br><span class="line">rm -rf 1000G_phase1.snps.high_confidence.hg38.ncbi.old.vcf</span><br></pre></td></tr></table></figure>

<p>再次运行，提示<code>success</code>。</p>
<h4 id="（二）"><a href="#（二）" class="headerlink" title="（二）"></a>（二）</h4><p>如果提示：<code>A USER ERROR has occurred: Input /home/yico/WES/gatk_resource_bundle/hg38/dbsnp_146.hg38.ncbi.vcf must support random access to enable queries by interval. If it&#39;s a file, please index it using the bundled tool IndexFeatureFile</code>或下载了原始的tbi索引，则需使用IndexFeatureFile对VCF文件重新建立索引。</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#</span><span class="bash">usage</span></span><br><span class="line">gatk IndexFeatureFile -I &lt;input.vcf&gt;  </span><br></pre></td></tr></table></figure>


    </div>

    
    
    

    <footer class="post-footer">

        

          <div class="post-nav">
            <div class="post-nav-item">
                <a href="/2022/01/17/WES03/" rel="prev" title="测序数据比对到参考基因组">
                  <i class="fa fa-chevron-left"></i> 测序数据比对到参考基因组
                </a>
            </div>
            <div class="post-nav-item">
                <a href="/2022/01/24/WES05/" rel="next" title="Variants Calling">
                  Variants Calling <i class="fa fa-chevron-right"></i>
                </a>
            </div>
          </div>
    </footer>
  </article>
</div>






</div>
  </main>

  <footer class="footer">
    <div class="footer-inner">


<div class="copyright">
  &copy; 
  <span itemprop="copyrightYear">2022</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Yico Yuan</span>
</div>
<div class="busuanzi-count">
    <span class="post-meta-item" id="busuanzi_container_site_uv">
      <span class="post-meta-item-icon">
        <i class="fa fa-user"></i>
      </span>
      <span class="site-uv" title="总访客量">
        <span id="busuanzi_value_site_uv"></span>
      </span>
    </span>
    <span class="post-meta-item" id="busuanzi_container_site_pv">
      <span class="post-meta-item-icon">
        <i class="fa fa-eye"></i>
      </span>
      <span class="site-pv" title="总访问量">
        <span id="busuanzi_value_site_pv"></span>
      </span>
    </span>
</div>

    </div>
  </footer>

  
  <script src="https://cdn.jsdelivr.net/npm/animejs@3.2.1/lib/anime.min.js" integrity="sha256-XL2inqUJaslATFnHdJOi9GfQ60on8Wx1C2H8DYiN1xY=" crossorigin="anonymous"></script>
<script src="/js/comments.js"></script><script src="/js/utils.js"></script><script src="/js/motion.js"></script><script src="/js/schemes/muse.js"></script><script src="/js/next-boot.js"></script>

  
<script src="https://cdn.jsdelivr.net/npm/hexo-generator-searchdb@1.4.0/dist/search.js" integrity="sha256-vXZMYLEqsROAXkEw93GGIvaB2ab+QW6w3+1ahD9nXXA=" crossorigin="anonymous"></script>
<script src="/js/third-party/search/local-search.js"></script>




  <script src="/js/third-party/pace.js"></script>

  
  <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>





</body>
</html>
